src/xrpld/app/tx/detail/DepositPreauth.cpp

```diff
@@ -18,15 +18,285 @@
 //==============================================================================
 
 #include <xrpld/app/tx/detail/DepositPreauth.h>
+#include <xrpld/app/tx/detail/OnChainCredentials.h>
 #include <xrpld/ledger/View.h>
 #include <xrpl/basics/Log.h>
 #include <xrpl/protocol/Feature.h>
 #include <xrpl/protocol/Indexes.h>
 #include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/digest.h>
 #include <xrpl/protocol/st.h>
 
+#include <optional>
+#include <unordered_set>
+
 namespace ripple {
 
+static std::optional<std::reference_wrapper<STArray const>>
+getOptArray(STTx const& tx, SField const& field)
+{
+    int index = tx.getFieldIndex(field);
+    if (index < 0)
+        return std::nullopt;
+    STBase const& f = tx.peekAtIndex(index);
+    if (f.getSType() != STI_ARRAY)
+        return std::nullopt;
+    return std::cref(static_cast<STArray const&>(f));
+}
+
+static AccountID const&
+getAcount(STObject const& o, SField const& field)
+{
+    static AccountID const emtpy;
+    int index = o.getFieldIndex(field);
+    if (index < 0)
+        return emtpy;
+    STBase const& f = o.peekAtIndex(index);
+    if (f.getSType() != STI_ACCOUNT)
+        return emtpy;
+    return static_cast<STAccount const&>(f).value();
+}
+
+static Slice
+getCredType(STObject const& o, SField const& field)
+{
+    static Slice const emtpy;
+    int index = o.getFieldIndex(field);
+    if (index < 0)
+        return emtpy;
+    STBase const& f = o.peekAtIndex(index);
+    if (f.getSType() != STI_VL)
+        return emtpy;
+    return static_cast<STBlob const&>(f).value();
+}
+
+static std::optional<std::reference_wrapper<STVector256 const>>
+getOptVector(STTx const& tx, SField const& field)
+{
+    int index = tx.getFieldIndex(field);
+    if (index < 0)
+        return std::nullopt;
+    STBase const& f = tx.peekAtIndex(index);
+    if (f.getSType() != STI_VECTOR256)
+        return std::nullopt;
+    return std::cref(static_cast<STVector256 const&>(f));
+}
+
+static uint256
+hashAuthCred(STObject const& o)
```

Comment:
Don't need this convenience function. This is `sha512Half(o.getAccountID(sfIssuer),
                   o.getFieldVL(sfCredentialType));`. Can just use it directly in the code instead.

Reply:
fixed